pkg/adapters/rules/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,284 @@
+			DeleteFunc: workqueue.EnqueueingDeleteFunc(adapter.deleteRules),
+		},
+	)
+	logger.Info("Calling Adapter start" )
+	adapter.workqueue.Start()
+
+	return adapter, adapter.Start()
+}
+
+// Start synchronizing the Kubernetes adapter.
+func (a *Adapter) Start() error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	if a.stopChan != nil {
+		err := fmt.Errorf("kubernetes adapter already started")
+
+		logger.WithError(err).Errorf("Failed starting Kubernetes adapter")
+		return err
+	}
+	a.stopChan = make(chan struct{})
+
+	go a.rulesController.Run(a.stopChan)
+	return nil
+}
+
+func createRulesClient(config Config) (*rest.RESTClient, error) {
+	var kubeConfig *rest.Config
+	if config.URL != "" {
+		kubeConfig = &rest.Config{
+			Host:        config.URL,
+			BearerToken: config.Token,
+		}
+	} else {
+		var err error
+		kubeConfig, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed loading Kubernetes credentials from service account")
+		}
+	}
+
+	groupversion := unversioned.GroupVersion{
+		Group:   ResourceGroupName,
+		Version: ResourceVersion,
+	}
+
+	kubeConfig.GroupVersion = &groupversion
+	kubeConfig.APIPath = "/apis"
+	kubeConfig.ContentType = runtime.ContentTypeJSON
+	kubeConfig.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: api.Codecs}
+
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			scheme.AddKnownTypes(
+				groupversion,
+				&RoutingRule{},
+				&RoutingRuleList{},
+				&api.ListOptions{},
+				&api.DeleteOptions{},
+			)
+			return nil
+		})
+	schemeBuilder.AddToScheme(api.Scheme)
+
+	client, err := rest.RESTClientFor(kubeConfig)
+	if err != nil {
+		logger.Printf("Failed to create the in-cluster client. %s\n", err)
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// Stop synchronizing the Kubernetes adapter.
+func (a *Adapter) Stop() error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	if a.stopChan == nil {
+		err := fmt.Errorf("kubernetes adapter not started")
+		logger.WithError(err).Errorf("Failed stopping Kubernetes adapter")
+		return err
+	}
+	close(a.stopChan)
+	a.stopChan = nil
+	a.workqueue.Stop()
+	return nil
+}
+
+// ListRules queries for the list of rules currently exist.
+func (a *Adapter) ListRules(f *a8api.RuleFilter) (*a8api.RulesSet, error) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	rules := make([]a8api.Rule, 0, len(a.rules))
+	for _, rule := range a.rules {
+		rules = append(rules, *rule)
+	}
+	rules = f.Apply(rules)
+	rulesSet := &a8api.RulesSet{
+		Rules:    rules,
+		Revision: 1, // TODO: how to fill revision ?
+
+	}
+	logger.Info("List rules: %v", rulesSet.Rules )
+	return rulesSet, nil
+}
+
+// addRule is the callback invoked by the Kubernetes cache when a rule API resource is added.
+func (a *Adapter) addRule(obj interface{}) {
+	rule, ok := obj.(*RoutingRule)
+	if !ok {
+		logger.Warnf("Invalid rule added: object is of type %T", obj)
+		return
+	}
+	if rule.Status.State == RuleStateInvalid {
```

Comment:
Instead of duplicating the logic across `addRule()` and `updateRules()`, add some `storeRule()` method and call it from both. Also note that switching on `rule.Status.State` may be a better fit here.